src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractBuildingDependentRequestResolver.java

```diff
@@ -0,0 +1,96 @@
+package com.minecolonies.coremod.colony.requestsystem.resolvers.core;
+
+import com.minecolonies.api.colony.requestsystem.location.ILocation;
+import com.minecolonies.api.colony.requestsystem.manager.IRequestManager;
+import com.minecolonies.api.colony.requestsystem.request.IRequest;
+import com.minecolonies.api.colony.requestsystem.requestable.IRequestable;
+import com.minecolonies.api.colony.requestsystem.requester.IRequester;
+import com.minecolonies.api.colony.requestsystem.token.IToken;
+import com.minecolonies.coremod.colony.buildings.AbstractBuilding;
+import com.minecolonies.coremod.colony.requestsystem.management.handlers.RequestHandler;
+import com.minecolonies.coremod.colony.requestsystem.requesters.BuildingBasedRequester;
+import com.minecolonies.coremod.colony.requestsystem.requesters.IBuildingBasedRequester;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Abstract class used as a base class for request resolver that handle request coming from the building they are originating from.
+ * @param <R> The type of request that they handle
+ */
+public abstract class AbstractBuildingDependentRequestResolver<R extends IRequestable> extends AbstractRequestResolver<R> implements IBuildingBasedRequester
+{
+
+    public AbstractBuildingDependentRequestResolver(
+      @NotNull final ILocation location,
+      @NotNull final IToken<?> token)
+    {
+        super(location, token);
+    }
+
+    @Override
+    public boolean canResolve(
+      @NotNull final IRequestManager manager, final IRequest<? extends R> requestToCheck)
+    {
+        if (!manager.getColony().getWorld().isRemote)
+        {
+            final ILocation requesterLocation = requestToCheck.getRequester().getRequesterLocation();
+            if (requesterLocation.equals(getRequesterLocation()))
+            {
+                final AbstractBuilding building = getBuilding(manager, requestToCheck.getToken()).map(r -> (AbstractBuilding) r).get();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Call "Optional#isPresent()" before accessing the value. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3655)